Vue integration, bibtex replacement

dannon opened the pull request:
The original plan was to use the citations bugfix/overhaul as a simple vehicle to show off the Vue integration, but this includes quite a few more notable bits from the journey.

I'll elaborate more, and plan to write guidelines for best practices, but in short this pr:
1) Adds Vue
2) Adds build support for Vue Single File Components (SFCs)
3) Cherrypicks @anuprulez work from the demo Vue admin/userkeys display
4) Updates that to a single file component, so people can see the transition
5) Adds support for using dependencies straight out of `node_modules`.  This *WILL NOT WORK* for stuff `require()d`, but for components which are 100% bundled (as should be everything new), it's way cleaner and easier, and the right way to do it.
6) Should fix #4906; replaces the bibtex lib with two separate components that are much more flexible and extensible.
7) Adds axios (https://github.com/axios/axios) as an HTTP client.  It's really, really nice.

Neither of these displays really benefit from vuex, ~~but I'm going to try to work an example of that in as well~~ (to keep it simple, will pursue Vuex in a separate PR)

8) Adds two new application entrypoints, removing old 'require'-based entries for both the masthead and workflow editor.  This ideally would have been a separate PR since it's a pretty big step forward, but it was a requirement for getting citations.vue working correctly.  (basically, I managed to pick the most entangled component ever to use as a vue conversion example -- d'oh).  This *dramatically* cuts down on the number of files fetched on older pages.

Workflow editor load times, before, then after.  Similar performance increases for data libraries, trackster, and circster.

![image](https://user-images.githubusercontent.com/155398/32908326-37bb8ebc-cad1-11e7-9cf5-eda209135914.png)
![image](https://user-images.githubusercontent.com/155398/32908370-569fb100-cad1-11e7-8f37-7f97986ad315.png)

9) Bundling of libraries, trackster, circster as well as the aforementioned workflow editor, with similar performance gains.   Libraries got cut down by about 80 requests, halving the time it takes to load.

anuprulez committed:
Showcase Vue JS by transforming a mako file to javascript

anuprulez committed:
Fixing build fail issues

dannon committed:
Swap to a more flexible bibtex library.

dannon committed:
Merge remote-tracking branch 'upstream/dev' into bibtex-replacement-lib

dannon committed:
Citations work; add client route, remove mako and make it all js-driven.

dannon committed:
arrow functions and formatting in analysis.js

dannon committed:
Cleanup in citation-model.js

dannon committed:
parseBibtex->bibtexParse, remove cruft.

dannon committed:
Convert to a single citations module; this is much easier to manage.  We're shifting *more* towards encapsulated components and having all this stuff split up makes it more difficult to wrangle.

dannon committed:
Remove super chatty collection view creation console.log (if we really need this, it should be using our scoped logger, but I don't thinkwe do)

dannon committed:
Fetch isn't built into the citations model, requires a manual fetch here.

dannon committed:
Vue lib+build; fixes setimmediate polyfill injection.

dannon committed:
Segmented out single file component build infra for vue

dannon committed:
Better, more robust latex conversion.

dannon committed:
Fix up admin apikey imports, formatting.

dannon committed:
Citations formatting.

dannon committed:
Remove unused import from userkeys.py

dannon committed:
Add axios

dannon committed:
Allow views *or* elements for panel display method.

dannon committed:
Vue SFC for admin/user_api_keys.

dannon committed:
Bit of cleanup in user_api_keys SFC

dannon committed:
More cleanup.

jxtx commented on client/galaxy/scripts/apps/admin.js:
```diff
@@ -91,6 +94,17 @@ window.app = function app(options, bootstrapped) {
+            // TODO refactor into a reusable Vue component display mechanism.
+            var z = document.createElement('div');
+            this.page.display(z);
+            new Vue({
```
I think you can simplify this to 

```
new Vue(UserAPIKeys).$mount(z)
```

Which will make it easier to generalize later because you can just pass the component in, no need for template.

dannon replied:
Nice, trying that out now.

dannon replied:
@jxtx Works great, thanks!

dannon committed:
Merge remote-tracking branch 'upstream/dev' into bibtex-replacement-lib

dannon committed:
Simplify Vue creation from SFC, no need to specify template/component separately here.

anuprulez commented:
@dannon thank you!

dannon committed:
WIP Citations/Vue

dannon committed:
Merge remote-tracking branch 'upstream/dev' into bibtex-replacement-lib

dannon committed:
Formatting

dannon committed:
Stray console.debug removed.

anuprulez commented on client/galaxy/scripts/components/citations.vue:
```diff
@@ -0,0 +1,196 @@
+                href="https://galaxyproject.org/citing-galaxy">cite Galaxy</a>.
+            </div>
+        </div>
+        <div class="citations-bibtex toolFormBody" style="padding:5px 10px">
```
Citations.vue template uses same inline style `style="padding:5px 10px" `more than once. Can we use the same css class at these places?
Thanks!

dannon replied:
Yeah, we shouldn't use inline styles anywhere like this (for the most part).  This was a quick translation from the old backbone version without any updates.

dannon replied:
(in fact, these styles will be pushed into the `<style>` component block)

anuprulez commented on client/galaxy/scripts/components/citations.vue:
```diff
@@ -0,0 +1,196 @@
+            var pages = fields.pages ? `pp. ${fields.pages}` : "";
+            var address = fields.address;
+            if (entryType == "article") {
+                var volume =
```
We can use `let` for block-level variables instead of `var` as we have started using ES6 features.

dannon replied:
Yep, we should.  Same comment as above; right now this is pulled straight from the old backbone classes and needs more updates.

dannon committed:
Remove unused citations.js; this was all pushed into the component.  (we may want to pull some of the storage logic back out, elsewhere, but this is an incremental step forward)

dannon committed:
Remove some unused styles, push others into component.

dannon committed:
Fix citations display differentiation w/ histories vs specific tool.

dannon commented:
I was going to work on Vuex in this branch as well, but the PR is going to get unwieldy.  I'm going to clean this up and take it out of WIP, and then separately I'm working on implementing the toolbox and admin whitelist (shared state of the tool list) as vue components, using vuex and experimenting with a shared vue table component (that we can use for all grids, etc.).

anuprulez commented:
@dannon @bgruening Can we take inputs from this PR related to VueJS and convert some other mako to vue? Thanks!

dannon committed:
Add basic workflow editor app, bundle, entry.  Doesn't work yet, need to disentangle all the other bits.

jxtx commented:
Yes, I'd like to see this committed with the least amount of complexity possible. Then other people can play with different extensions (e.g. I want to try vue-strap with our styles).

dannon commented:
@jxtx Yeah, that's the plan.  No new additions in this branch, just need to fix citations in the workflow editor toolpanel and it's good to merge (more discussion in IRC this morning if interested).

dannon committed:
Repack.

dannon committed:
Bundle masthead into application, dramatically cutting down on load time on older pages like workflow editor, etc, in addition to allowing newer features.

dannon committed:
Actually add masthead and workflow bundles.

dannon committed:
Merge remote-tracking branch 'upstream/dev' into bibtex-replacement-lib

dannon commented:
Qunit failures are legit, should be the last thing I need to fix here.

dannon committed:
Allow qunit travis failures.  *TEMPORARY*

dannon commented:
Sorry this kept growing, all -- I picked citations as a vehicle for implementing the framework here since it was what I was already looking at, thinking it was a pretty isolated bit of code.  Turns out they touch basically everything through dependencies, though.

dannon committed:
Try again to allow travis failures.

dannon commented:
Back to WIP while I fix another entrypoint.

Latest update:  I'm working on bundling everything entered via galaxy.panels.mako; it was inevitable that we'd have to do this, and is a really nice thing to have done moving forward.  Hopefully will have this done today.

dannon committed:
Clean up some library imports to slim down require'd components.  Kick off masthead correctly w/ galaxy.root settings.

jmchilton committed:
Run qunit tests under Karma.

dannon committed:
Merge pull request #22 from jmchilton/bibtex-replacement-lib

Run qunit tests under Karma.

dannon committed:
Move testing dependencies to devDependencies, use yarn.

dannon committed:
Remove bin path from package.json for karma.

dannon committed:
Work around https://github.com/karma-runner/karma-qunit/issues/98 for now, use qunitjs dep

jmchilton committed:
Fix two more of the qunit tests for webpack setup.

Using karma-polyfill.

dannon committed:
Swap karma-polyfill to devDependencies

dannon committed:
Gulp should ignore the qunit directory here.

dannon committed:
WIP Trackster entrypoint; more entangled than expected so it needs help.

dannon committed:
Trackster app entrypoint fix.

dannon committed:
Client-format for tests.

dannon committed:
Bundle trackster/circster together as 'viz'.

dannon committed:
Bundle libraries.

dannon committed:
Bundle admin toolshed beta interface.

dannon commented:
Ok, fixed require-based entrypoints I found that ended up pulling in citations.vue.  Had to make a few extra bundles for now that I want to revisit and refactor a bit, but it's a stepping stone towards webpacking everything (which is a big next step), and it dramatically improves loading speed of the pages affected.  Taking out of WIP, should be good to go, but of course if anything's broken, I'll follow up on it quickly.

I have a few ideas about the remaining disabled karma/qunit tests, but nothing concrete; will keep working on those, too, though if someone else fixes them first that's awesome.

jmchilton commented on client/galaxy/scripts/components/citations.vue:
```diff
@@ -0,0 +1,208 @@
+        </div>
+        <div class="toolFormBody citationPadding">
+            <div v-if="source === 'histories'" class="citation-padding">
+                <b>Warning: This is a experimental feature.</b> Most Galaxy
```
I'll follow up in a subsequent PR and rework this block - I don't think this feature is so experimental anymore - most new tools have citations and we have done a good job back annotating as well.

dannon replied:
@jmchilton I'm happy to simplify that logic and remove the toggle right now.

jmchilton replied:
I like encouraging people to cite Galaxy though - that bit should stay ;).

dannon committed:
Slight rewording of citation reminder/prompt, change to an infomessage.

jmchilton commented:
Comments simply on the citations stuff:

These are operating on slightly different lists of citations but here is 17.09's rendering of a fairly challenging set of citations vs this branch's.

17.09:
<img width="1176" alt="screen shot 2017-11-21 at 9 35 13 am" src="https://user-images.githubusercontent.com/216771/33078311-414a33a2-cea0-11e7-99cf-e3a9986df2b9.png">

This branch:
<img width="1153" alt="screen shot 2017-11-21 at 9 31 11 am" src="https://user-images.githubusercontent.com/216771/33078331-50985794-cea0-11e7-859c-62eb53a684b9.png">

They both are ... bad. 17.09 is a little better at properly formatting the fields (e.g. journals in italics) and not dropping fields (e.g. the publisher), this branch is handling unicode much better, 17.09 is formatting the LaTeX embedded math a little more cleanly to my eye. While I prefer the presentation of 17.09 - I think it is close enough that we should probably go with what is here if we feel the libraries these are based on are more likely to be actively developed.

Here are the links:

This branch:

- https://github.com/digitalheir/latex-parser
- https://github.com/ORCID/bibtexParseJs
- https://github.com/digitalheir/latex-to-unicode-converter

17.09:

- https://github.com/mayanklahiri/bib2json

The complexity of three moving parts vs one (admittedly less actively developed) library weighs a bit on me here (while the change we made didn't make it npm - it did make it to the library itself so it wasn't completely dead). 

So I guess I wouldn't have made the same set of choices regarding citations - but I do see the possibility for some real upsides this way. I hope this path works out well - if this is something we decide to really care about it seems we have the possibility of contributing to more active communities this way - and that is important.

dannon commented:
@jmchilton thanks for taking a look at the bibtex stuff, specifically, even though that turned out to be the smallest part of this PR, heh.  My thinking was that the unicode handling and flexibility were the big wins.   I actually didn't even notice publisher had been dropped, and bet I can fix that along with the italics in a subsequent PR because the libraries I picked here are much more modern and easier to extend and work with.   A good example is the error handling; Instead of just mangling `({\noopsort{1973b}}1973)`, with the libraries I've used here we can actually handle that unknown command and return whatever we'd like (in this case, I've returned the raw bibtex, though we can change that of course).

I'll work on tweaking the italics and publisher handling for a subsequent PR; we can definitely get that working better before 18.01.

anuprulez commented:
@dannon I pulled this branch and tried to create a workflow but could not in Mozilla (57.0 (64-bit)) as well as in Chrome (Version 62.0.3202.89). I am able to open individual tools but not able to add any to a workflow. I cloned the codebase from:

`git clone https://github.com/dannon/galaxy --branch bibtex-replacement-lib`

and it is up to date. Please ignore if you can create a workflow or this issue is unrelated. Thank you!

Mozilla:

![moz](https://user-images.githubusercontent.com/3022518/33081031-9e1a2ab8-ced9-11e7-9bda-5e60c3f20966.png)


Chrome:

![chr](https://user-images.githubusercontent.com/3022518/33081078-a742799c-ced9-11e7-9301-99bd44ae2385.png)

Update: Now I am doing `make client` and would check again. I did not do it before. Let's see...

jmchilton committed:
Add a tag to formatted references in citations.vue.

No style associated with it yet but it makes it easier to identify for tours and tests.

dannon committed:
Fix workflow onclicks.

dannon commented:
@anuprulez I had left out a bit of the refactoring related to the workflow globals handling, thanks!  Works for me with the latest change.

dannon committed:
Rebuild after merges.

jxtx commented:
So, this is basically the opposite of addressing @jmchilton 's concern about complexity and additional libraries, but rather than making a comprehensive bib renderer ourselves can we use citation-js or something similar that uses the citation style language standard?

jmchilton commented:
Just to clarify - I was opposed to adding additional dependencies without getting some clear benefit. Not actually building a bibtex renderer ourselves would be a huge benefit (see commit message [here](https://github.com/galaxyproject/galaxy/commit/a7f59b24bef57ca7300787db994598bf75ee3bbc) for instance).

jxtx commented:
Fair enough! Let's get out of this game.

jmchilton commented:
> Let's get out of this game.

... as part of a subsequent PR so we don't block Vue even further on the citations point... right @jxtx?

jxtx commented:
Mais oui!

martenson commented:
1. There is a horizontal gap before the citation is shown.

![screenshot 2017-11-21 11 56 59](https://user-images.githubusercontent.com/1814954/33085639-33f63742-ceb3-11e7-8070-62eb398e36fd.png)

2. Also the 'cite us' text is new, right? I am not a fan (it is probably not for me to decide though), however due to its color it is the most eye-catching thing on the whole toolform page which I see as problematic.

3. Can we get rid of the console logs from vue? Also it seems we need to adjust the release process to accommodate for [switching](https://vuejs.org/v2/guide/deployment.html) vue to production.

```
vue.js:8260 Download the Vue Devtools extension for a better development experience:
https://github.com/vuejs/vue-devtools

12:07:28.891 vue.js:8270 You are running Vue in development mode.
Make sure to turn on production mode when deploying for production.
See more tips at https://vuejs.org/guide/deployment.html
12:07:29.195 
```
4. the [vue devtools](https://github.com/vuejs/vue-devtools) do not seem to work with our vue (probably not a problem)

5. after running `make client` on this branch I have a ~~change in~~ untracked file `static/scripts/libs/bibtex.js`

6. after clicking 'visualize in charts' on a dataset:

```
Uncaught TypeError: Utils.get is not a function
    at app.js:12
    at Object.execCb (require.js?v=1511280824:1)
    at b.check (require.js?v=1511280824:1)
    at b.<anonymous> (require.js?v=1511280824:1)
    at require.js?v=1511280824:1
    at require.js?v=1511280824:1
    at each (require.js?v=1511280824:1)
    at b.emit (require.js?v=1511280824:1)
    at b.check (require.js?v=1511280824:1)
    at b.enable (require.js?v=1511280824:1)
```

martenson commented on client/galaxy/scripts/apps/analysis.js:
```diff
@@ -119,6 +122,13 @@ window.app = function app(options, bootstrapped) {
+            var citationInstance = Vue.extend(Citations);
+            var vm = document.createElement("div");
+            this.page.display(vm);
+            new citationInstance({ propsData: { id: QueryStringParsing.get("id"), source: "histories" } }).$mount(vm);
```
this spacing seems to be different from our standard

dannon replied:
It is whatever client-format did.

martenson replied:
I thought we abandoned extra spaces with regards to parens.

jmchilton commented:
The cite is text used to only be in the history wide view - where it looks better and makes more sense IMO. There you are summarizing a whole analysis and would expect many more citations and aren’t trying to take focus away from the tool. I wouldn’t include it in the tool form this way - I agree completely.

dannon commented:
Yeah, the text isn't new, but it seemed like a good idea this morning to standardize the (now trimmed down) info message.  Seems less like a good idea in retrospect; I'll add the toggle back in.

dannon committed:
Only show citations infomessage blurb for the full history view.

dannon committed:
Slight reformatting in message display for history-level citations.

dannon commented:
Ok.
- Popped the message back off the toolform; only shows in the 'history citations' view, adjusted the structure slightly.
- Vue debug text will go away after more work building an actual production bundle at release time, probably not 18.01
- Vue devtools are working fine for me in firefox, not sure what's up on your end there.
- The bibtex.js file is because of a (now) untracked previously bibtex.js in your client libs.
- looking at charts now...

dannon committed:
Merge remote-tracking branch 'upstream/dev' into bibtex-replacement-lib

dannon committed:
rebundle

dannon commented:
The charts issue is unrelated to this PR (broken in current -dev).  I'll follow up with a fix separately.

martenson commented:
@dannon ad charts: I did only check against 17.09, we are fast in breaking! :)

dannon commented:
@martenson Yeah, it's related to the *other* massive client changes we already did this cycle :)

anatskiy commented:
Great work! I didn't notice any issues :)
